csrc/device_lower/pass/allocation.cpp

```diff
@@ -174,6 +175,74 @@ class AllocationInserter : public kir::ExprMutator {
 
     info.allocation_domains = std::make_unique<std::vector<IterDomain*>>();
 
+    // TODO: Today, we always allocate loop domain, even if the allocation
+    // domain is explicitly set. This is clearly not the right thing to do,
+    // and we should fix this in the future. However, today, we still don't
+    // have a clear design of how to allocate tensors with explicit allocation
+    // domain. This problem is very difficult to solve, and there are many
+    // things to consider. For example, if the allocation domain contains a
+    // subset of inlined loop IDs, we should not allocate the inlined IDs.
+    // But what if the opposite is true? What if the allocation domain
+    // does not contain all inlined IDs? Is this considered an error, or
+    // a valid case that we need to infer which to allocate from the loop
+    // domain? We need to think about this carefully and come up with a
+    // clear design. For now, we just allocate the loop domain for historical
+    // reasons for all cases except for the Hopper MMA output tensor.
+    //
+    // Hopper MMA output tensor is a special case because the loop domain
+    // is scheduled in a way that the entire tile is parallelized by MMA, and
+    // The TIDx parallelization is a new broadcast dimension that is not
+    // connected to any other IterDomains. This way of scheduling effectively
+    // makes the loop domain 128x larger than the allocation domain, because
+    // the allocation domain is sharded on threads but the loop domain is not.
+    if ((info.buffer->definition()->isA<MmaOp>() &&
+         isHopper(info.buffer->definition()->as<MmaOp>()->macro()))) {
+      std::unordered_set<IterDomain*> exclude_ca_ids;
+      for (auto i : c10::irange(info.alloc_pos)) {
+        auto ca_id = info.buffer->axis(i);
+        if (!ir_utils::isMemorySharedAcross(
+                info.buffer->getMemoryType(), ca_id->getParallelType())) {
+          exclude_ca_ids.insert(ca_id);
+        }
+      }
+
+      const std::vector<IterDomain*>& domain_to_alloc =
+          info.buffer->hasAllocation() ? info.buffer->getAllocationDomain()
+                                       : info.buffer->getLoopDomain();
+
+      for (auto id : domain_to_alloc) {
+        if (exclude_ca_ids.find(id) == exclude_ca_ids.end()) {
+          // Don't use reduction/stride/broadcast/device axis in the
+          // allocation computation
+          if (id->isReduction() || id->isStride() || id->isBroadcast() ||
+              id->isDeviceDim()) {
+            continue;
+          }
+          if (ir_utils::isMemoryPartitionedAcross(
+                  info.buffer->getMemoryType(), id->getParallelType())) {
+            continue;
+          }
+          info.allocation_domains->push_back(id);
+          if (gpu_lower->caMap()->idExistsInMap(id, IdMappingMode::LOOP)) {
+            id = gpu_lower->caMap()->getConcreteMappedID(
```

Comment:
Shouldn't this use the loop promotion map?

Reply:
I believe we should. This part of code is copy-pasted from below without much thinking.😅

Reply:
fixed